### PR TITLE
[Reshard] reshard from partial to replicated in same placement

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR types

New features
### PR changes

Others
### Description

Pcard-73145

Reshard方面：

- 新增从partial到replicated的转换，通过直接调用all_reduce通信得到全量的数据。

丰富语义：

- Reshard以设备为中心，在设备间完成重切分，目前从tensor视角描述的TensorDistAttr存在表达能力不足的问题。为此，提供tensor到设备视角的转换，方便reshard对当前状态进行判断。

### CODE REVIEW

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_reshard_p_to_r
                       PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE" TIMEOUT 100)`

File: paddle/phi/core/distributed/auto_parallel/dist_attr.cc [hunk 1]
```diff
@@ -348,5 +348,69 @@ bool TensorDistAttr::empty() const {
...
+    }
+  }
+  for (auto& itr : partial_status_) {
+    PADDLE_ENFORCE_LT(
+        itr.first,
+        ndim,
+        errors::InvalidArgument(
+            "Partial axis %ld can not exceed the ndim of process_mesh %ld",
+            itr.first,
+            ndim));
```
Comment (JZ-LIANG) [inline paddle/phi/core/distributed/auto_parallel/dist_attr.cc:375]: could be better to double check here that the dim to partial is not be used as sharded.

File: paddle/phi/core/distributed/auto_parallel/dist_attr.cc [hunk 2]
```diff
@@ -348,5 +348,69 @@ bool TensorDistAttr::empty() const {
...
+  for (size_t i = 0; i < dims_mapping_.size(); ++i) {
+    if (dims_mapping_[i] != -1) {
+      PADDLE_ENFORCE_LT(
+          dims_mapping_[i],
+          ndim,
+          errors::InvalidArgument(
+              "Split axis %ld can not exceed the ndim of process_mesh %ld",
+              dims_mapping_[i],
+              ndim));
+      placement[dims_mapping_[i]] = std::make_shared<ShardStatus>(i);
```
Comment (JZ-LIANG) [inline paddle/phi/core/distributed/auto_parallel/dist_attr.cc:365]: why shared_ptr ？ the return placement is supposed for temporary usage and should not be hold by others.

**Review by JZ-LIANG — APPROVED** [review]
LGTM and might revise in next PR

**Pull request merged**